### ISSUE TITLE
[GR-40463] Add initial support for remote management over JMX in Native Images

### ISSUE THREAD

**Issue by roberttoyonaga** [oracle/graal#4732]
### TL;DR
Currently connections via JMX are not supported in native images. This work is to add initial support for that.

### Goals
Achieve similar functionality as in OpenJDK where remote management of managed beans is possible. JMX support will be optionally added at image build time through `--enable-montioring=jmxclient,jmxserver` options.

### Details
Building a JMX server or client should be possible. Connection should work with authentication as well as SSL. When a JMX server is built and run with remote JMX options, the server should start up right away at runtime similar to in OpenJDK. Configuration will be done in separate client and server features for modularity.

**Comment by christianhaeubl** [comment]
FYI @fniephaus, @christianwimmer, @vjovanov

**Comment by fniephaus** [comment]
Thanks for volunteering to work on this, @roberttoyonaga! Let us know if you need any help from our side.

**Comment by roberttoyonaga** [comment]
The Javadoc [here](https://github.com/oracle/graal/blob/a96025039f6c18d1d4031240c5ea4741356a5852/substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/management/ManagementSupport.java#L77-L78) says that the MBeanServer should be initialized at runtime.  Is the intent here that the user, in their application code, manually initializes it, if required, by calling `ManagementFactory.getPlatformMBeanServer()`?  Would it be better if the `MBeanServer` and  `JMXConnectorServer` were started automatically via a `jdk.RuntimeSupport.Hook` startup-hook if a flag was set at native image build time (similar to openjdk)?

**Comment by roberttoyonaga** [comment]
It looks like [jdk.internal.agent.Agent](https://github.com/openjdk/jdk17/blob/master/src/jdk.management.agent/share/classes/jdk/internal/agent/Agent.java#L61) initializes the `JMXConnecterServer`. It also initializes the `MBeanServer` by calling `ManagementFactory.getPlatformMBeanServer()`. Hotspot calls on this class to do all the initialization if JMX is specified at runtime.

Maybe I can reuse this class in a start up hook in the `ManagementFeature` to automatically initialize JMX services at runtime if the user specifies they want JMX functionality.

**Comment by roberttoyonaga** [comment]
After some basic experimenting, it seems like reusing the JDK infrastructure is possible. And it is also possible to do the `MBeanServer` and `JMXConnecterServer` initialization at runtime with a start-up hook so that a client is able to connect. However, reusing the JDK infrastructure requires configuration for things like reflection, serialization, proxy classes, and resource bundles to be specified at image build time. Currently, I have everything specified in json files created by running the native image tracing agent. This means that anyone wanting to use JMX would need the proper configuration at image build time. Is this an acceptable solution? Could there be a better way?

**Comment by fniephaus** [comment]
Cool, that's good news! For JDK code, we do use internal features instead of JSON config files. An example is the [`JDKInitializationFeature`](https://github.com/oracle/graal/blob/292f4553a9cf5c341fff9f9503647f9d98dc36f2/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/jdk/JDKInitializationFeature.java), but we also have more specific ones such as [`JNIRegistrationAwt`](https://github.com/oracle/graal/blob/master/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/jdk/JNIRegistrationAwt.java).

How large are those config files? Could you maybe turn them into something like a `JMXInitializationFeature`? What do you suggest, @christianhaeubl?

**Comment by roberttoyonaga** [comment]
Thanks Fabio! Yes I can give that a try. The reflection json file is about 600 lines (so maybe ~200 classes), but all the other config files are pretty small (<50 entries).

**Comment by roberttoyonaga** [comment]
Doing all the configuration inside of a feature is working well. I am thinking of splitting the configuration for the client and server into 2 separate `Feature` classes. What do you think of this idea? 

Do we still need a separate [ManagementFeature](https://github.com/oracle/graal/blob/2e526905b95c861b6eb0b2d1528d139f2979d7b0/substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/management/ManagementFeature.java#L59)  or can this code be merged with the JMX server configuration feature code? 

I am running into a runtime linking error with native JNI method `jdk.internal.agent.FileSystemImpl.isAccessUserOnly0(String)` I have tried adding the library with `NativeLibraries.addStaticJniLibrary(String, String ...)` but the liking error persists. Do you know  I can resolve this?

**Comment by fniephaus** [comment]
Splitting the config into server and client features sounds reasonable. If you think it can be merged with the ManagementFeature, please go ahead. This is something we can still discuss during code review.

I'm not sure what the error is about. Maybe @christianhaeubl has an idea? Probably best you open a PR with your changes so that we can reproduce this and start commenting on the implementation.

**Comment by roberttoyonaga** [comment]
Ok I have created a draft PR here: https://github.com/oracle/graal/pull/4786
I added some comments to explain the current state of the work.

**Comment by fniephaus** [comment]
Thanks! What do we need to do to reproduce the linking issue you mentioned?

**Comment by roberttoyonaga** [comment]
To reproduce the error you need to start any native image executable built with my branch using these flags: `-Dcom.sun.management.jmxremote.authenticate=false -Dcom.sun.management.jmxremote.ssl=false -Dcom.sun.management.jmxremote.local.only=false -Dcom.sun.management.jmxremote.port=9999 `.  This will enable jmx password auth. The JMX server should start automatically and fail with a linking error.

**Comment by roberttoyonaga** [comment]
> To reproduce the error you need to start any native image executable built with my branch using these runtime flags: `-Dcom.sun.management.jmxremote.authenticate=true -Dcom.sun.management.jmxremote.ssl=false -Dcom.sun.management.jmxremote.local.only=false -Dcom.sun.management.jmxremote.port=9999 -Dcom.sun.management.jmxremote.access.file=jmxremote/jmxremote.access -Dcom.sun.management.jmxremote.password.file=jmxremote/jmxremote.password `. This will enable jmx password auth. The JMX server should start automatically and fail with a linking error. I've checked-in the dummy password files in `jmxremote` directory as well.

The linking issue has been resolved.

**Comment by fniephaus** [comment]
Great! What was the fix?

**Comment by roberttoyonaga** [comment]
I was linking the libmanagement_agent native library using NativeLibraries.addStaticJniLibrary(String) but its methods were not being recognized. I needed to use `PlatformNativeLibrarySupport.addBuiltinPkgNativePrefix(String)` to essentially register the prefix `jdk_internal_agent`. This way the native methods with this prefix inside this library could be used properly.

**Comment by roberttoyonaga** [comment]
I think the first iteration of the work is ready for review now. The associated PR is updated. Please let me know if the current approach is on the right track.

**Comment by fniephaus** [comment]
Thanks, @roberttoyonaga! We're going to take a look asap :)

**Comment by enderwsp** [comment]
Is there a development version available to verify jmx to facilitate ，further verification of tomcatmbean thread pool information
